### ISSUE TITLE
Modifying an element makes a chord symbol move

### ISSUE THREAD

**Issue by BanjoJake** [Jojo-Schmitz/MuseScore#344]
In the attached file, delete, edit, or move the subtitle, and the G chord in bar 5 shifts downward.
Un-do does not move it back, but saving the file twice does.  
I believe this is related to  #340

Observed in
Windows 10, MuseScore version (64-bit): 3.7.0.8035433412

[Changing Element Moves Chord.zip](https://github.com/Jojo-Schmitz/MuseScore/files/14397364/Changing.Element.Moves.Chord.zip)

Edit (2024-06-04): 
Confirmed bug in 3.7.0.9312831347 / 4d876fe with Automatic Placement = ON for all chord symbols
Confirmed bug not in 4.3.1-241490900/ 026c26b

**Comment by Jojo-Schmitz** [comment]
It is different from #340 in that it does happen in 3.6.2 too, but in 3.6.2 it moves all chord symbols down, not just thao single one.
Indeed doesn't happen in 4.2.1 though

**Comment by Jojo-Schmitz** [comment]
Strange, now I can't reproduce it in 3.6.2 anymore

**Comment by BanjoJake** [comment]
Confirmed doesn't happen in 4.2.1 nor 3.6.2

**Comment by BanjoJake** [comment]
May have found a work-around, or at least a clue - select all chords, turn Automatic Placement OFF, then back on.

**Comment by worldwideweary** [comment]
This doesn't happen for me in my personal build. Dusting off some old changes I had something similar happen to me when I was building a polychord entry system (the stacked type with line), and I made this change here:
```cpp
 void Harmony::layout1()
       {
       if (isLayoutInvalid())
             createLayout();
       if (textBlockList().empty())
             textBlockList().append(TextBlock());
-      auto positionPoint = calculateBoundingRect();    // for normal symbols this is called in layout: computeMinWidth()
+      calculateBoundingRect(); // for normal symbols this is called in layout: computeMinWidth()
       if (hasFrame())
             layoutFrame();
       score()->addRefresh(canvasBoundingRect());
-      setPos(positionPoint);
```

@Jojo-Schmitz  If this code seems familiar, we went over it when updating the TextFrames of chord symbol.... this is just another step afterwards. 

Testing of this change would be welcome. I guess I can try a quick PR

**Comment by Jojo-Schmitz** [comment]
Let's try: #349

**Comment by Jojo-Schmitz** [comment]
@BanjoJake mind to test?

**Comment by BanjoJake** [comment]
Yup, should be able to start testing in about an hour.    Thanks!!JakeSent from my iPhoneOn Feb 29, 2024, at 8:53 AM, Joachim Schmitz ***@***.***> wrote:﻿
@BanjoJake mind to test?

**Comment by worldwideweary** [comment]
If we're lucky, this can be closed. If an exception is found, definitely revisit the issue tracker ;)

**Comment by BanjoJake** [comment]
Thanks, but I found another example where a chord symbol moves incorrectly in v3.7 and v3.6.2:
In the attached example, I move the first Dm Chord in measure 9 up slightly, so it's roughly even with the second chord.  
It then jumps up a bit higher, so I have to move it down slightly.  
Then I do File-Save and it jumps back down to where it was originally.

Observed on Windows 10 in MuseScore (64-bit) 3.6.2.548021803, 3.7.0.8127680949, and several earlier versions of 3.7.  
I'll install and test v4.2.1 shortly.

[Chord Moves on Save.zip](https://github.com/Jojo-Schmitz/MuseScore/files/14473748/Chord.Moves.on.Save.zip)

**Comment by BanjoJake** [comment]
And another:  Moving an element incorrectly moves a distant Chord Symbol:   
In the attached example, I move the first Dm Chord in measure 9 up slightly, so it's roughly even with the second chord.  
It then jumps up a bit higher, so I have to move it down slightly.   Then I move the last chord symbol (E) in measure 4 (or the last triplet bracket in measure 8), and the first Dm Chord in measure 9 jumps back down to where it was originally.

[Moving an element incorrectly moves a distant Chord Symbol.zip](https://github.com/Jojo-Schmitz/MuseScore/files/14473891/Moving.an.element.incorrectly.moves.a.distant.Chord.Symbol.zip)

Observed on Windows 10 in MuseScore (64-bit) 3.6.2.548021803 and 3.7.0.8127680949.  
Testing on v4.2.1 is inconclusive so far - observed once, could not repro.

**Comment by Jojo-Schmitz** [comment]
>  and several earlier versions of 3.7.

Still the case?

**Comment by BanjoJake** [comment]
Not after applying the 'work-around': Select all chords, turn Automatic Placement OFF, then back ON, then Save.   Now none of the objectionable behavior occurs in any version.  So maybe it's just a matter of 'resetting' the Automatic Placement of Chord Symbols ?

**Comment by worldwideweary** [comment]
For what it's worth, the chord mentioned in the examples has a minimum distance of -999. B.S. can happen when something like that exists not in accord with the rest of the symbols. I'd reset all minimum distance with Select All Similar first (not as a workaround, but in good taste), before doing any manual offsetting.

**Comment by Jojo-Schmitz** [comment]
B.S. -> Bachelor of Science?    ;-)

**Comment by worldwideweary** [comment]
Bickering Stuntmen o_O

**Comment by BanjoJake** [comment]
@worldwideweary - yes, now finding these sprinkled liberally throughout my scores.   Selecting all, then Reset to Style Default en masse did not seem to work, seems I have to do them individually.  May be easier to edit the mscx as text.    And I'm wodering if this setting was masked by #340 and #344 ?     Anyway, MANY THANKS for finding this !!

**Comment by worldwideweary** [comment]
@BanjoJake: I'd also try [Reset Shapes and Positions] if "Reset to Style" didn't work, which I think is Ctrl+R by default. Seems to work here. . . shouldn't have to do it individually

**Comment by BanjoJake** [comment]
> For what it's worth, the chord mentioned in the examples has a minimum distance of -999.
> ... now finding these sprinkled liberally throughout my scores.  

FYI, found an old post in the forum that Musescore sets the minimum distance to -999 if an item is moved too close to the staff.  Since then I have replicated this behavior which is intentional per design.    That explains where they all came from.

**Comment by Jojo-Schmitz** [comment]
The fix for this issue caused #489, so I'me reverting that fix and reopening this issue.

**Comment by BanjoJake** [comment]
Observing in 3.7.0.9312831347:  Adjusting a chord symbol vertically sometimes makes other chord symbols jump up or down seemingly randomly, often changing the minimum distance to -999.   Similarly saving the file will sometimes make a chord symbol jump up or down.   Please let me know if more examples would be helpful.

**Comment by BanjoJake** [comment]
For now I am reverting to 3.7.0.8981527544 for regular use, but will be happy to test any new version upon request.   Also: confirmed that Studio v4.3.1-241490900 behaves like 3.7.0.9312831347: Adjusting a chord symbol vertically sometimes makes other chord symbols jump up or down seemingly randomly.  I don't see minimum distance in the 4.3.1 properties window, so unable to report on that aspect.   

Incidentally, it seems that in 4.3.1, clicking on certain chord symbols fails to select them.  This behavior seems random, and I don't have time to test it further right now.

**Comment by Jojo-Schmitz** [comment]
The build numbers don't help much, the commit hash from Help > About MuseScore much more

**Comment by BanjoJake** [comment]
> The build numbers don't help much, the commit hash from Help > About MuseScore much more

Sorry, I thought it was the other way around !   Editing my previous comment to include that info (but leaving in the build numbers for my own reference).

**Comment by BanjoJake** [comment]
So far, it seems that always setting **all** chord symbols to have Automatic Placement = OFF is a reasonable work-around.    I suppose in debugging this, the first stage would be with Automatic Placement = ON for all chord symbols, then a mix ?

**Comment by worldwideweary** [comment]
I think I already know what the problem is, but it's screwed up how to fix it. It has to do with how the systems get collected, because within that functionality are the layout calls to [measure bases] which in turn perform layouts to all elements, including harmony texts.

When a user edits/moves a harmony text, system collection occurs for some reason (almost seems superfluous but there are reasons due to automatic placement etc) it invokes the default layout (y offset = 0).... and then it finalizes by "systemic layout" (second pass) afterwards... but it only does this to the finalized system, the "actual" system on the page. 

The algorithm goes through just the actual system though when 'figuring out' how big the 'actual' system will be, and while it does that, it calls the damned *layouts* via measure base when checking to see if it fits (hence the layout y offset drop only happens on the first measure of the next system)... and we're left with a zeroed y-offset since the "left over" measure bases that don't make it don't get the entire "system layout" afterwards.

The question is how to make a solution that makes sense

**Comment by BanjoJake** [comment]
Is it relevant that the bug does not occur in Mu4 ?  (just tested again in 4.3.1-241490900/ https://github.com/Jojo-Schmitz/MuseScore/commit/026c26b874add66eefdc6d099564dcf3437b78a6

**Comment by worldwideweary** [comment]
I've got a quick test PR. It's attempting to do what was done earlier, but only applied at the createShapes position which happens all the time when editing/moving a harmony text. That's to say, the vertical leveling thing still works but this time doesn't reset the layout when creating shapes which happens each time an edit/move occurs.

Please test it as thoroughly as possible to see if something wrong happens somewhere.... would be cool if it worked. And it's just a one liner so if needs to be reverted then it's no problem. Figured worth testing rather than doing nothing for now.

**Comment by BanjoJake** [comment]
The behavior of  **Changing Element Moves Chord.mscz**  is different in  3.7.0.9376084770 / 09bbacc - The buggy behavior still occurs, but takes more moving elements around before it happens.   Still experimenting ...

**Comment by worldwideweary** [comment]
Thanks for experimenting. I'll check the file also and try some things

**Comment by worldwideweary** [comment]
@BanjoJake Precise steps to manifest the problem with the "changing element moves chord" file would definitely help. I'm not getting the G chord to move after moving/editing  texts and harmony around. Let me know if you can "recreate it" on demand precisely. Thanks!

**Comment by BanjoJake** [comment]
Still struggling with this, as the behavior is much improved.  Basically, after moving the subtitle down and up, then alternately repeatedly moving the C chord in bar 6 up & down, and moving the G chord in bar 5 up & down, at some point the C chord jumps when you move the G chord.  But I'm still trying to generate precise step-by-step instructions ...

**Comment by BanjoJake** [comment]
Ignore the instructions in the subtitle.   Move the C chord in bar 6 up (eg, to -4.53sp) & down, but still **well above the staff line** (eg, -1.41sp), the minimum distance changes to -999, resulting in the reported behavior.

This may be a different / additional bug.  Still testing.

**Comment by BanjoJake** [comment]
Here we go:     Ignore the instructions in the subtitle.   Change the Bottom gap of the top vertical frame to 8.00sp.  Move the G chord in bar 1 up (eg -6.53sp) and down (eg -2.13sp).    Then move the G chord in bar 5 up (eg -6.07sp).  The G chord in bar 1 jumps down.   Minimum distance does not change to -999 for either chord this time, so maybe two bugs going on here ...

**Comment by worldwideweary** [comment]
Ok, verified the "Here we go" comment. It happens in 3.6.2 also (G chord in bar one jumping down) so it's not a result of this particular change. Thanks! What I'm mainly worried about is having a negative side effect from this particular PR. So far so good it seems to me. Not sure what's causing the problem you've pointed out though. 

One thing I've noticed is that some of this seems correlated with the V-Align properties.  Like for example (i put on the bounding boxes to see it easier):

[vokoscreenNG-2024-06-05_20-57-01.webm](https://github.com/Jojo-Schmitz/MuseScore/assets/7139517/d8c24d83-b9e8-449a-9404-f298b3c1cd7b)

**Comment by Jojo-Schmitz** [comment]
That PR is not yet merged, so not in the latest 3.x nightly build.
Should it get merged is it ready?

**Comment by worldwideweary** [comment]
Just changed the code to do a check if bbox was empty or not... should be okay so far so good but "you never know"

**Comment by BanjoJake** [comment]
OK, confirmed that this is fixed in MuseScore (64-bit): 3.7.0.9413109854 / e51f231 (Windows 10), ie, the behaviorin  [Changing Element Moves Chord.zip](https://github.com/Jojo-Schmitz/MuseScore/files/14397364/Changing.Element.Moves.Chord.zip) and https://github.com/Jojo-Schmitz/MuseScore/files/14473891/Moving.an.element.incorrectly.moves.a.distant.Chord.Symbol.zip   (https://github.com/Jojo-Schmitz/MuseScore/files/14473748/Chord.Moves.on.Save.zip was an artifact of minimum distance = -999 per worldwideweary)
